### PR TITLE
Fill ENTSOE category extension from CGMES generating unit description

### LINKED ISSUE #2103: [CGMES] Fill ENTSOE category extension
* **Do you want to request a *feature* or report a *bug*?**

Feature.

* **What is the current behavior?**

We have a `GeneratorEntsoeCategory` extension, but it's not created when importing a CGMES file.
The information is lost.

* **What is the expected behavior?**

We should retrieve that information and create the extension to expose it to the user.

* **What is the motivation / use case for changing the behavior?**

pypowsybl users expected to have that information filled in, when importing a CGMES file.

@zamarrenolm : would you know how to retrieve that information ?

### PR DESCRIPTION
**Pull request by zamarrenolm**
Signed-off-by: Luma <zamarrenolm@aia.es>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
Closes #2103

### CODE REVIEW

**Comment by zamarrenolm** [comment]
Reviewed impact on CGMES import performance: total import time of 25 TYNDP test cases increased from 43.5s to 45s in average (10 samples)

**Review by sylvlecl — APPROVED** [review]

**Review by geofjamg — APPROVED** [review]

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/EntsoeCategoryPostProcessor.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            String generatingUnitId = sm.getId("GeneratingUnit");
+            if (generatingUnitId != null) {
+                String description = sm.getId("generatingUnitDescription");
+                // String contains only digits
+                if (StringUtils.isNumeric(description)) {
+                    String generatorId = sm.getId("SynchronousMachine");
+                    Generator g = network.getGenerator(generatorId);
+                    if (g != null) {
+                        try {
+                            int code = Integer.parseInt(description);
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/EntsoeCategoryPostProcessor.java]: thanks!

**Review by annetill — CHANGES_REQUESTED** [review]
Code equal to zero is allowed indeed...

**Comment by geofjamg** [comment]
> Code equal to zero is allowed indeed...


    private static int checkCode(int code) {
        if (code < 1) {
            throw new IllegalArgumentException("Bad generator ENTSO-E code " + code);
        }
        return code;
    }

**Comment by geofjamg** [comment]
Looks like it is forbidden in extension

**Comment by annetill** [comment]
> Looks like it is forbidden in extension

Euh... Fabrice told me that 0 and 99 are used for HVDC and SVC :-) So we have to change the extension...

**Comment by geofjamg** [comment]
> > Looks like it is forbidden in extension
> 
> Euh... Fabrice told me that 0 and 99 are used for HVDC and SVC :-) So we have to change the extension...

It works with actual TYNDP? CVG -> IIDM create all theses extensions without any error? So there is no need.

**Comment by annetill** [comment]
> > > Looks like it is forbidden in extension
> > 
> > 
> > Euh... Fabrice told me that 0 and 99 are used for HVDC and SVC :-) So we have to change the extension...
> 
> It works with actual TYNDP? CVG -> IIDM create all theses extensions without any error? So there is no need.

I think that we can merge the PR like this. If zero as a code is needed, it will be in a dedicated PR as it will change the extension implementation. Okay?

**Comment by zamarrenolm** [comment]
> > > > Looks like it is forbidden in extension
> > > 
> > > 
> > > Euh... Fabrice told me that 0 and 99 are used for HVDC and SVC :-) So we have to change the extension...
> > 
> > 
> > It works with actual TYNDP? CVG -> IIDM create all theses extensions without any error? So there is no need.
> 
> I think that we can merge the PR like this. If zero as a code is needed, it will be in a dedicated PR as it will change the extension implementation. Okay?

Ok for me. I added the check > 0 to avoid creating an invalid extension (the exception was captured anyway, so the import was not finished with error).

**Pull request merged**